pkg/traceql/engine_metrics.go

```diff
@@ -1060,51 +1129,76 @@ func (b *SimpleAdditionAggregator) Combine(in []*tempopb.TimeSeries) {
 				Values:    make([]float64, b.len),
 				Exemplars: make([]Exemplar, 0, len(ts.Exemplars)),
 			}
+			if b.initWithNaN {
+				for i := range existing.Values {
+					existing.Values[i] = nan
+				}
+			}
+
 			b.ss[ts.PromLabels] = existing
 		}
 
-		for _, sample := range ts.Samples {
-			j := IntervalOfMs(sample.TimestampMs, b.start, b.end, b.step)
-			if j >= 0 && j < len(existing.Values) {
-				existing.Values[j] += sample.Value
-			}
-		}
+		b.aggregationFunc(b.start, b.end, b.step, &existing, ts.Samples)
+		b.aggregateExemplars(ts, &existing)
 
-		for _, exemplar := range ts.Exemplars {
-			if b.exemplarBuckets.testTotal() {
-				break
-			}
-			interval := IntervalOfMs(exemplar.TimestampMs, b.start, b.end, b.step)
-			if b.exemplarBuckets.addAndTest(interval) {
-				continue // Skip this exemplar and continue, next exemplar might fit in a different bucket
-			}
+		b.ss[ts.PromLabels] = existing
+	}
+}
 
-			labels := make(Labels, 0, len(exemplar.Labels))
-			for _, l := range exemplar.Labels {
-				labels = append(labels, Label{
-					Name:  l.Key,
-					Value: StaticFromAnyValue(l.Value),
-				})
-			}
-			value := exemplar.Value
-			if math.IsNaN(value) {
-				value = 0 // TODO: Use the value of the series at the same timestamp
-			}
-			existing.Exemplars = append(existing.Exemplars, Exemplar{
-				Labels:      labels,
-				Value:       value,
-				TimestampMs: uint64(exemplar.TimestampMs),
+func (b *SimpleAggregator) aggregateExemplars(ts *tempopb.TimeSeries, existing *TimeSeries) {
+	for _, exemplar := range ts.Exemplars {
+		if b.exemplarBuckets.testTotal() {
+			break
+		}
+		interval := IntervalOfMs(exemplar.TimestampMs, b.start, b.end, b.step)
+		if b.exemplarBuckets.addAndTest(interval) {
+			continue // Skip this exemplar and continue, next exemplar might fit in a different bucket	}
+		}
+		labels := make(Labels, 0, len(exemplar.Labels))
+		for _, l := range exemplar.Labels {
+			labels = append(labels, Label{
+				Name:  l.Key,
+				Value: StaticFromAnyValue(l.Value),
 			})
 		}
-
-		b.ss[ts.PromLabels] = existing
+		value := exemplar.Value
+		if math.IsNaN(value) {
+			value = 0 // TODO: Use the value of the series at the same timestamp
+		}
+		existing.Exemplars = append(existing.Exemplars, Exemplar{
+			Labels:      labels,
+			Value:       value,
+			TimestampMs: uint64(exemplar.TimestampMs),
+		})
 	}
 }
 
-func (b *SimpleAdditionAggregator) Results() SeriesSet {
+func (b *SimpleAggregator) Results() SeriesSet {
 	return b.ss
 }
 
+// Simple addition aggregator. It adds existing values with the new sample. New indicates if the existing values were just initialized, ugly
+func simpleAdditionAggregator(start, end, step uint64, existing *TimeSeries, samples []tempopb.Sample) {
+	for _, sample := range samples {
+		j := IntervalOfMs(sample.TimestampMs, start, end, step)
+		if j >= 0 && j < len(existing.Values) {
+			existing.Values[j] += sample.Value
```

Comment:
I kind of want to thin these out to just the 1 or 2 lines in question, so that the interval check/calculation isn't duplicated, and put it back in Combine.  Something like:

```go
func (existing, new float64) float64

add := func(...) { return existing + new }
min := func(...) { if math.IsNaN(existing) || ... { return new } else { return existing } }

// in Combine
for _, sample := range samples {
	j := IntervalOfMs(sample.TimestampMs, start, end, step)
	if j >= 0 && j < len(existing.Values) {
		existing.Values[j] = fn(existing.Values[j], sample.Value)
```

Reply:
That's fine although I'm not sure it will work for the avg_over_time since we need to know the number of non-nan elements for the given interval to compute it

Reply:
but we can fine-tune it later

Reply:
It would be nice if Go had the  "=>" operator

add := (a float64, b float64) => a + b

like in c#

```cs
Func<string> greet = () => "Hello, World!";
Console.WriteLine(greet());
```

Reply:
infering the return type ...

Reply:
Yes I was thinking about average as well, since it requires state.  That is true and a single func pointer isn't enough.